server/src/pam_listener.rs

```diff
@@ -0,0 +1,191 @@
+    /// Current user's secret, used to unlock their keyring <username, secret>
+    user_secrets: Arc<RwLock<std::collections::HashMap<String, Secret>>>,
+}
+
+impl PamListener {
+    pub fn new(service: Service) -> Self {
+        let socket_path = std::env::var("OO7_PAM_SOCKET")
+            .map(PathBuf::from)
+            .unwrap_or_else(|_| PathBuf::from(DEFAULT_PAM_SOCKET_PATH));
+
+        Self {
+            socket_path,
+            service,
+            user_secrets: Arc::new(RwLock::new(std::collections::HashMap::new())),
+        }
+    }
+
+    /// Start the PAM listener
+    pub async fn start(self) -> Result<(), Error> {
+        if let Some(parent) = self.socket_path.parent() {
+            tokio::fs::create_dir_all(parent).await?;
+        }
+
+        if self.socket_path.exists() {
+            tokio::fs::remove_file(&self.socket_path).await?;
+        }
+
+        let listener = UnixListener::bind(&self.socket_path)?;
+
+        tracing::info!("PAM listener started on {}", self.socket_path.display());
+
+        let perms = std::fs::Permissions::from_mode(0o600);
+        std::fs::set_permissions(&self.socket_path, perms)?;
+
+        let listener = Arc::new(listener);
+
+        // Accept connections in a loop
+        loop {
+            match listener.accept().await {
+                Ok((stream, _addr)) => {
+                    let pam_listener = self.clone();
+
+                    tokio::spawn(async move {
+                        if let Err(e) = pam_listener.handle_connection(stream).await {
+                            tracing::error!("Error handling PAM connection: {}", e);
+                        }
+                    });
+                }
+                Err(e) => {
+                    tracing::error!("Error accepting PAM connection: {}", e);
+                }
+            }
+        }
+    }
+
+    /// Handle a single PAM connection
+    async fn handle_connection(&self, mut stream: UnixStream) -> Result<(), Error> {
+        tracing::debug!("Accepted PAM connection");
+
+        // Read the message length prefix (4 bytes, little-endian)
+        let mut length_bytes = [0u8; 4];
+        stream.read_exact(&mut length_bytes).await?;
+        let message_length = u32::from_le_bytes(length_bytes) as usize;
+
+        let mut message_bytes = vec![0u8; message_length];
+        stream.read_exact(&mut message_bytes).await?;
+
+        let message = PamMessage::from_bytes(&message_bytes)
+            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e.to_string()))?;
+
+        tracing::info!("Received PAM authentication for user: {}", message.username);
+
+        let secret = Secret::from(message.secret);
+        self.user_secrets
+            .write()
+            .await
+            .insert(message.username.clone(), secret.clone());
+
+        // Try to unlock the default collection with this secret
+        match self.try_unlock_collections(&secret).await {
+            Ok(_) => {
```

Comment:
Does it make sense to keep listening to the socket after we already unlocked the keyring?

Reply:
My initial thought was to keep listening in case the user logs out and then logs in again or other user authentication operations that might trigger PAM. It is okay to keep around i would say